internal/lang/marks/marks.go

```diff
@@ -45,3 +50,146 @@ const Sensitive = valueMark("Sensitive")
 // another value's type. This is part of the implementation of the console-only
 // `type` function.
 const TypeType = valueMark("TypeType")
+
+type DeprecationCause struct {
+	By      addrs.Referenceable
+	Message string
+}
+
+type deprecationMark struct {
+	Cause DeprecationCause
+}
+
+func (m deprecationMark) GoString() string {
+	return "marks.Deprecated"
+}
+
+// Deprecated marks a given value as deprecated with specified DeprecationCause.
+func Deprecated(v cty.Value, cause DeprecationCause) cty.Value {
+	for m := range v.Marks() {
+		dm, ok := m.(deprecationMark)
+		if !ok {
+			continue
+		}
+
+		// Already marked as deprecated for this cause.
+		if addrs.Equivalent(dm.Cause.By, cause.By) {
+			return v
+		}
+	}
+
+	return v.Mark(deprecationMark{
+		Cause: cause,
+	})
+}
+
+// DeprecatedOutput marks a given values as deprecated constructing a DeprecationCause
+// from module output specific data.
+func DeprecatedOutput(v cty.Value, addr addrs.AbsOutputValue, msg string) cty.Value {
+	_, callOutAddr := addr.ModuleCallOutput()
+	return Deprecated(v, DeprecationCause{
+		By:      callOutAddr,
+		Message: msg,
+	})
+}
+
+// DeprecatedDiagnosticsInBody composes deprecation diagnostics based on deprecation marks inside
+// the cty.Value. It uses hcl.Body to properly reference deprecated attributes in final diagnostics.
+func DeprecatedDiagnosticsInBody(val cty.Value, body hcl.Body) (cty.Value, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+
+	val, deprecatedPathMarks := unmarkDeepWithPathsDeprecated(val)
+
+	for _, pm := range deprecatedPathMarks {
+		for m := range pm.Marks {
+			cause := m.(deprecationMark).Cause
+			diags = diags.Append(tfdiags.AttributeValue(
+				tfdiags.Warning,
+				"Value derived from a deprecated source",
+				fmt.Sprintf("This value is derived from %v, which is deprecated with the following message:\n\n%s", cause.By, cause.Message),
+				pm.Path,
+			))
+		}
+	}
+
+	return val, diags.InConfigBody(body, "")
+}
+
+// DeprecatedDiagnosticsInExpr composes deprecation diagnostics based on deprecation marks inside
+// the cty.Value. It uses hcl.Expression to properly reference deprecated attributes in final diagnostics.
+func DeprecatedDiagnosticsInExpr(val cty.Value, expr hcl.Expression) (cty.Value, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+
+	val, deprecatedPathMarks := unmarkDeepWithPathsDeprecated(val)
+
+	// Locate deprecationMarks and filter them out
+	for _, pm := range deprecatedPathMarks {
+		for m := range pm.Marks {
+			attr := tfdiags.FormatCtyPath(pm.Path)
+			// FormatCtyPath call could result in ".fieldA.fieldB" in some
+			// cases, so we want to remove the first dot for a friendlier message.
+			if len(attr) > 1 && attr[0] == '.' {
+				attr = attr[1:]
+			}
```

Comment:
FWIW we're using `tfdiags.FormatCtyPath` in the same way elsewhere with the intention that the leading dot makes it clearer that we're talking about a relative attribute path rather than an absolute one. I don't think removing it's necessarily bad in this case, but the inconsistency of doing it in only this one place (and thus making this particular message describe it differently than others) does give me some pause.

If we do decide to keep this, I'd note that the logic here is effectively the same as `strings.TrimPrefix(attr, ".")`, so might be clearer to write it that way.